### ISSUE TITLE
SHM and SRE should not be allowed to share a name

### ISSUE THREAD

**Issue by jemrobinson** [alan-turing-institute/data-safe-haven#1850]
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates.
- [x] This is a problem observed when **deploying** a Data Safe Haven.
- [x] I can reproduce this with the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :computer: System information

- **Operating System:** macOS
- **Data Safe Haven version:** develop @ c1d0adc

## :no_entry_sign: Describe the problem

If an SHM and SRE share a name then the keys in `pulumi.yaml` will be clobbered.

This happens because the YAML file does not distinguish between SHM and SRE projects, e.g.

```
projects:
  apple:
    encrypted_key: <key>
    stack_config:
      azure-native:location: uksouth
      ...
  green:
    encrypted_key: <key>
    stack_config:
      azure-native:location: uksouth
      ...
```

## :deciduous_tree: Log messages

<details>
<summary>Relevant log messages</summary>

```none
Your log details here
```

</details>

## :recycle: To reproduce

- Deploy an SHM
- Attempt to deploy an SRE with the same name

**Comment by JimMadge** [comment]
Does it let you do that? I thought I had prevented that from happening,

https://github.com/alan-turing-institute/data-safe-haven/blob/c1d0adc779d6c58b6901feea7a6ec1ca2ca27111/data_safe_haven/config/pulumi.py#L41-L54

Do we want to allow different components to have identical names?

**Comment by jemrobinson** [comment]
1. You can upload a config which contains an SHM and SRE with the same name.
2. If you run `dsh deploy sre <same name as shm>` then the deployment starts running.

It's possible that writing `pulumi.yaml` to Azure after the SRE has been deployed might fail, but I think we should be catching this at (1), not allowing the deployment to go ahead.

Alternatively, we could add an `shm-` or `sre-` prefix to the name in the `pulumi.yaml` file if you prefer?

**Comment by JimMadge** [comment]
Hmm, In that case I would guess 'create_or_select_project` would fetch the existing Pulumi configuration and stack state. Deploying would presumably then delete the SHM as it isn't defined in the SRE program.

I think adding an enum to `DSHPulumiProject` which is `SHM | SRE` would be a good solution. File/project/stack names could be modified based on that.

**Comment by JimMadge** [comment]
@jemrobinson Does this still need to be done, or are we expecting to remove the SHM entirely for rc2?

**Comment by jemrobinson** [comment]
Let's leave it until we've worked out whether that's possible.

**Comment by JimMadge** [comment]
Related to #1900. Will not be necessary if that is closed.

**Comment by JimMadge** [comment]
@jemrobinson is this fixed by #1943 ?

**Comment by jemrobinson** [comment]
No, it's still possible for an SHM and SRE to share a name. If so, this will cause a clash in the DSHPulumiConfig
